src/hardware_manager.rs

```diff
@@ -0,0 +1,332 @@
+
+            let reading = navigator_rs::SensorData {
+                adc: {
+                    if refresh_interval_us < 10000 {
+                        navigator_rs::ADCData { channel: [0.0; 4] }
+                    } else {
+                        lock.navigator.read_adc_all()
+                    }
+                },
+                temperature: lock.navigator.read_temperature(),
+                pressure: lock.navigator.read_pressure(),
+                accelerometer: lock.navigator.read_accel(),
+                magnetometer: lock.navigator.read_mag(),
+                gyro: lock.navigator.read_gyro(),
+            };
+
+            drop(lock);
+
+            *DATA.write().unwrap() = Data { state: reading };
+
+            let time_elapsed = time_start.elapsed().as_micros() as u64;
+
+            if time_elapsed > (refresh_interval_us * 2) {
+                log::info!("Monitor: Something went wrong, measurements not concluded with reading interval {refresh_interval_us} us, time elapsed: {time_elapsed} us");
+                NavigationManager::websocket_broadcast();
+                continue;
+            }
+
+            NavigationManager::websocket_broadcast();
+
+            let wait = refresh_interval_us as i64 - time_elapsed as i64;
+
+            if wait > 0 {
+                thread::sleep(std::time::Duration::from_micros(wait as u64));
+            }
+        }
+    }
+
+    fn data_logger(refresh_interval: u64, directory: String, filename: String) {
+        let file_path = format!("{}/{}", directory, filename);
+        let mut logger = DataLogger::new(&file_path).expect("Failed to create/open CSV file");
+
+        // Just let monitor run before
+        thread::sleep(std::time::Duration::from_millis(500));
+
+        log::info!("Datalogger started");
+
+        loop {
+            let reading = DATA.read().unwrap().state;
+
+            logger.log_data(&reading).expect("Failed to log data");
+
+            thread::sleep(std::time::Duration::from_micros(refresh_interval));
+        }
+    }
+
+    fn websocket_broadcast() {
+        // This package is broadcasted when it's created
+        let _package: crate::server::protocols::v1::structures::AnsPackage =
+            packages::reading(packages::Sensors::All, true);
+    }
+}
+
+// Help with conversion from navigator enum API to our stable API
+// impl_from_enum!(AdcChannel, navigator_rs::AdcChannel, Ch0, Ch1, Ch2, Ch3);
+impl_from_enum!(
+    PwmChannel,
+    navigator_rs::PwmChannel,
+    Ch1,
+    Ch2,
+    Ch3,
+    Ch4,
+    Ch5,
+    Ch6,
+    Ch7,
+    Ch8,
+    Ch9,
+    Ch10,
+    Ch11,
+    Ch12,
+    Ch13,
+    Ch14,
+    Ch15,
+    Ch16,
+    All
+);
+// impl_from_enum!(UserLed, navigator_rs::UserLed, Led1, Led2, Led3);
+
+impl From<navigator_rs::AxisData> for AxisData {
+    fn from(read_axis: navigator_rs::AxisData) -> Self {
+        Self {
+            x: read_axis.x,
+            y: read_axis.y,
+            z: read_axis.z,
+        }
+    }
+}
```

Comment:
You should check, but I believe that:
```suggestion
impl From<&navigator_rs::AxisData> for AxisData {
    fn from(read_axis: &navigator_rs::AxisData) -> Self {
        Self {
            x: read_axis.x,
            y: read_axis.y,
            z: read_axis.z,
        }
    }
}
```

should work

Reply:
not possible, need change something on navigator-rs